File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/handler/websocket/DisconnectRequestHandler.java
package net.prostars.messagesystem.handler.websocket;

import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.constant.MessageType;
import net.prostars.messagesystem.constant.UserConnectionStatus;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.DisconnectRequest;
import net.prostars.messagesystem.dto.websocket.outbound.DisconnectResponse;
import net.prostars.messagesystem.dto.websocket.outbound.ErrorResponse;
import net.prostars.messagesystem.service.ClientNotificationService;
import net.prostars.messagesystem.service.UserConnectionService;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class DisconnectRequestHandler implements BaseRequestHandler<DisconnectRequest> {

  private final UserConnectionService userConnectionService;
  private final ClientNotificationService clientNotificationService;

  public DisconnectRequestHandler(
      UserConnectionService userConnectionService,
      ClientNotificationService clientNotificationService) {
    this.userConnectionService = userConnectionService;
    this.clientNotificationService = clientNotificationService;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, DisconnectRequest request) {
    UserId senderUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());
    Pair<Boolean, String> result =
        userConnectionService.disconnect(senderUserId, request.getUsername());
    if (result.getFirst()) {
      clientNotificationService.sendMessage(
          senderSession,
          senderUserId,
          new DisconnectResponse(request.getUsername(), UserConnectionStatus.DISCONNECTED));
    } else {
      String errorMessage = result.getSecond();
      clientNotificationService.sendMessage(
          senderSession,
          senderUserId,
          new ErrorResponse(MessageType.DISCONNECT_REQUEST, errorMessage));
    }
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/DisconnectRequestHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.constant.MessageType
import com.yscorp.withpush.messagesystem.constant.UserConnectionStatus
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.DisconnectRequest
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.DisconnectResponse
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.ErrorResponse
import com.yscorp.withpush.messagesystem.service.ClientNotificationService
import com.yscorp.withpush.messagesystem.service.UserConnectionService
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession

@Component
@Suppress("unused")
class DisconnectRequestHandler(
    private val userConnectionService: UserConnectionService,
    private val clientNotificationService: ClientNotificationService
) : BaseRequestHandler<DisconnectRequest> {

    override fun handleRequest(senderSession: WebSocketSession, request: DisconnectRequest) {
        val senderUserId: UserId = senderSession.attributes[IdKey.USER_ID.value] as UserId
        val result: Pair<Boolean, String> =
            userConnectionService.disconnect(senderUserId, request.username)
        if (result.first) {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                DisconnectResponse(request.username, UserConnectionStatus.DISCONNECTED)
            )
        } else {
            val errorMessage = result.second
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(MessageType.DISCONNECT_REQUEST, errorMessage)
            )
        }
    }
}
